zebra-chain/src/primitives/redpallas/batch.rs

```diff
@@ -0,0 +1,265 @@
+// -*- mode: rust; -*-
+//
+// This file is part of redpallas.
+// Copyright (c) 2019-2021 Zcash Foundation
+// See LICENSE for licensing information.
+//
+// Authors:
+// - Deirdre Connolly <deirdre@zfnd.org>
+// - Henry de Valence <hdevalence@hdevalence.ca>
+
+//! Performs batch RedPallas signature verification.
+//!
+//! Batch verification asks whether *all* signatures in some set are valid,
+//! rather than asking whether *each* of them is valid. This allows sharing
+//! computations among all signature verifications, performing less work overall
+//! at the cost of higher latency (the entire batch must complete), complexity of
+//! caller code (which must assemble a batch of signatures across work-items),
+//! and loss of the ability to easily pinpoint failing signatures.
+//!
+
+use std::convert::TryFrom;
+
+use group::{Group, GroupEncoding};
+use halo2::arithmetic::FieldExt;
+use rand_core::{CryptoRng, RngCore};
+
+use super::{private::Sealed, scalar_mul::VartimeMultiscalarMul, *};
+
+// Shim to generate a random 128bit value in a [u64; 4], without
+// importing `rand`.
+fn gen_128_bits<R: RngCore + CryptoRng>(mut rng: R) -> [u64; 4] {
```

Comment:
Let's use doc comments here, and explain that some bytes are zero.

```suggestion
/// Shim to generate a random 128 bit value in a `[u64; 4]`, without
/// importing `rand`.
///
/// The final 128 bits are zero.
fn gen_128_bits<R: RngCore + CryptoRng>(mut rng: R) -> [u64; 4] {
```